# Cleaning data - sharks dataframe

# Paso 1: importar dataframe y hacer una primera observación de los datos que la componen

In [1125]:
import pandas as pd
import os
from re import sub
import numpy
from statistics import mean
%matplotlib inline
import matplotlib.pyplot as plt

En primer lugar, importo el dataframe con el que voy a trabajar para poder empezar a trabajar con los datos.

In [1126]:
data = pd.read_csv('attacks.csv', encoding='latin-1')
data.head()

Case Number         Date    Year        Type    Country             Area  \
0  2018.06.25  25-Jun-2018  2018.0     Boating        USA       California   
1  2018.06.18  18-Jun-2018  2018.0  Unprovoked        USA          Georgia   
2  2018.06.09  09-Jun-2018  2018.0     Invalid        USA           Hawaii   
3  2018.06.08  08-Jun-2018  2018.0  Unprovoked  AUSTRALIA  New South Wales   
4  2018.06.04  04-Jun-2018  2018.0    Provoked     MEXICO           Colima   

                         Location     Activity             Name Sex   ...  \
0     Oceanside, San Diego County     Paddling      Julie Wolfe    F  ...   
1  St. Simon Island, Glynn County     Standing  Adyson McNeely     F  ...   
2                    Habush, Oahu      Surfing      John Denges    M  ...   
3              Arrawarra Headland      Surfing             male    M  ...   
4                        La Ticla  Free diving   Gustavo Ramos     M  ...   

          Species           Investigator or Source                       pdf  \
0      White shark                R. Collier, GSAF      2018.06.25-Wolfe.pdf   
1              NaN  K.McMurray, TrackingSharks.com    2018.06.18-McNeely.pdf   
2              NaN  K.McMurray, TrackingSharks.com     2018.06.09-Denges.pdf   
3        2 m shark                  B. Myatt, GSAF  2018.06.08-Arrawarra.pdf   
4  Tiger shark, 3m                       A .Kipper      2018.06.04-Ramos.pdf   

                                        href formula  \
0  http://sharkattackfile.net/spreadsheets/pdf_di...   
1  http://sharkattackfile.net/spreadsheets/pdf_di...   
2  http://sharkattackfile.net/spreadsheets/pdf_di...   
3  http://sharkattackfile.net/spreadsheets/pdf_di...   
4  http://sharkattackfile.net/spreadsheets/pdf_di...   

                                                href Case Number.1  \
0  http://sharkattackfile.net/spreadsheets/pdf_di...    2018.06.25   
1  http://sharkattackfile.net/spreadsheets/pdf_di...    2018.06.18   
2  http://sharkattackfile.net/spreadsheets/pdf_di...    2018.06.09   
3  http://sharkattackfile.net/spreadsheets/pdf_di...    2018.06.08   
4  http://sharkattackfile.net/spreadsheets/pdf_di...    2018.06.04   

  Case Number.2 original order Unnamed: 22 Unnamed: 23  
0    2018.06.25         6303.0         NaN         NaN  
1    2018.06.18         6302.0         NaN         NaN  
2    2018.06.09         6301.0         NaN         NaN  
3    2018.06.08         6300.0         NaN         NaN  
4    2018.06.04         6299.0         NaN         NaN  

[5 rows x 24 columns]

Después de haber importado este dataframe, busco resolver algunas primeras dudas (¿Qué columnas la componen?, ¿Cuántos datos continenen?, ¿Cuántos nulos hay por columna?, etc.)

In [1127]:
#Número de filas y columnas
data.shape

(25723, 24)

In [1128]:
#Nombre de las columnas
data.columns

Index(['Case Number', 'Date', 'Year', 'Type', 'Country', 'Area', 'Location',
       'Activity', 'Name', 'Sex ', 'Age', 'Injury', 'Fatal (Y/N)', 'Time',
       'Species ', 'Investigator or Source', 'pdf', 'href formula', 'href',
       'Case Number.1', 'Case Number.2', 'original order', 'Unnamed: 22',
       'Unnamed: 23'],
      dtype='object')

In [1129]:
#tipo de datos
data.dtypes.head()

Case Number     object
Date            object
Year           float64
Type            object
Country         object
dtype: object

In [1130]:
#También me interesa saber cuantos nulos tengo en cada variable
data.isnull().sum().head()

Case Number    17021
Date           19421
Year           19423
Type           19425
Country        19471
dtype: int64

Creo una copia de la base de datos original, para trabajar desde esa copia (tiburon):

In [1131]:
tiburon = data
tiburon.head()

Case Number         Date    Year        Type    Country             Area  \
0  2018.06.25  25-Jun-2018  2018.0     Boating        USA       California   
1  2018.06.18  18-Jun-2018  2018.0  Unprovoked        USA          Georgia   
2  2018.06.09  09-Jun-2018  2018.0     Invalid        USA           Hawaii   
3  2018.06.08  08-Jun-2018  2018.0  Unprovoked  AUSTRALIA  New South Wales   
4  2018.06.04  04-Jun-2018  2018.0    Provoked     MEXICO           Colima   

                         Location     Activity             Name Sex   ...  \
0     Oceanside, San Diego County     Paddling      Julie Wolfe    F  ...   
1  St. Simon Island, Glynn County     Standing  Adyson McNeely     F  ...   
2                    Habush, Oahu      Surfing      John Denges    M  ...   
3              Arrawarra Headland      Surfing             male    M  ...   
4                        La Ticla  Free diving   Gustavo Ramos     M  ...   

          Species           Investigator or Source                       pdf  \
0      White shark                R. Collier, GSAF      2018.06.25-Wolfe.pdf   
1              NaN  K.McMurray, TrackingSharks.com    2018.06.18-McNeely.pdf   
2              NaN  K.McMurray, TrackingSharks.com     2018.06.09-Denges.pdf   
3        2 m shark                  B. Myatt, GSAF  2018.06.08-Arrawarra.pdf   
4  Tiger shark, 3m                       A .Kipper      2018.06.04-Ramos.pdf   

                                        href formula  \
0  http://sharkattackfile.net/spreadsheets/pdf_di...   
1  http://sharkattackfile.net/spreadsheets/pdf_di...   
2  http://sharkattackfile.net/spreadsheets/pdf_di...   
3  http://sharkattackfile.net/spreadsheets/pdf_di...   
4  http://sharkattackfile.net/spreadsheets/pdf_di...   

                                                href Case Number.1  \
0  http://sharkattackfile.net/spreadsheets/pdf_di...    2018.06.25   
1  http://sharkattackfile.net/spreadsheets/pdf_di...    2018.06.18   
2  http://sharkattackfile.net/spreadsheets/pdf_di...    2018.06.09   
3  http://sharkattackfile.net/spreadsheets/pdf_di...    2018.06.08   
4  http://sharkattackfile.net/spreadsheets/pdf_di...    2018.06.04   

  Case Number.2 original order Unnamed: 22 Unnamed: 23  
0    2018.06.25         6303.0         NaN         NaN  
1    2018.06.18         6302.0         NaN         NaN  
2    2018.06.09         6301.0         NaN         NaN  
3    2018.06.08         6300.0         NaN         NaN  
4    2018.06.04         6299.0         NaN         NaN  

[5 rows x 24 columns]

# Hipótesis


Después de haber observado el DataFrame y el tipo de información que la integra, considero que sería interesante realizar mi trabajo con respecto a la siguiente hipótesis: En los últimos años el número de ataques de tiburón tienen consecuencias menos importantes para sus víctimas. Compararé los años 2000-2010 y los años 1950-1960, observando el número total de ataques fatales así como su proporción con respecto al total de ataques.

# Paso 2: Hacer la limpieza de datos

Aunque existen muchas maneras de limpiar datos, he decidido quedarme exclusivamente con la información que me permite realizar el trabajo de forma directa. He tomado esta decisión ya que considero que la recogida de datos contiene muchos errores y tiene poca fiabilidad. Por esta razón, me parece peligroso basarme en otras columnas que no sean aquellas que necesito directamente para mi trabajo. Veremos hasta qué punto los resultados que voy a obtener son significativos, con respecto a la muestra con la que voy a operar.

In [1132]:
#Elimino las columnas con las que no voy a operar, creando la variable col_malas

col_malas = tiburon.drop(columns = ['Date','Type','Sex ','Country','Area','Location','Activity','Name','Age','Injury','Time','Species ','Investigator or Source','pdf','href formula','href','Case Number.1','Case Number.2','original order','Unnamed: 22','Unnamed: 23'], axis=1)
tiburon = col_malas
tiburon.head()



Case Number    Year Fatal (Y/N)
0  2018.06.25  2018.0           N
1  2018.06.18  2018.0           N
2  2018.06.09  2018.0           N
3  2018.06.08  2018.0           N
4  2018.06.04  2018.0           N

No es necesario cambiar el nombre de nuestras columnas ya que no tienen errores (contrariamente al nombre de otras columnas que hemos borrado).

Para entrenarme a crear funciones, y sabiendo que voy a tener que limpiar muchos datos, he decidido crear una función para conocer número de valores de todas las columnas:

In [1133]:
def cls_counts(df):
    for col in df.columns:
        print("---%s---" %col)
        print(df[col].value_counts())

In [1134]:
#Pruebo mi función:
cls_counts(tiburon)

---Case Number---
0                 2400
2009.12.18           2
2013.10.05           2
1907.10.16.R         2
1980.07.00           2
1913.08.27.R         2
1920.00.00.b         2
2014.08.02           2
2005.04.06           2
1923.00.00.a         2
1990.05.10           2
2006.09.02           2
1966.12.26           2
1915.07.06.a.R       2
2012.09.02.b         2
1962.06.11.b         2
1983.06.15           2
1946.00.00.a         1
2017.08.29           1
1965.12.19           1
1878.09.02.a         1
1950.04.09.b         1
1987.06.13           1
1960.08.10           1
2016.04.09           1
1965.12.10           1
1996.01.12           1
2000.06.09.a         1
2014.08.29.b         1
1958.12.13           1
                  ... 
1919.03.16           1
1994.00.00.a         1
2018.03.09.a         1
1973.11.25           1
1939.10.04           1
1963.04.22           1
1948.11.18           1
1936.04.22           1
2001.08.25           1
1901.00.00           1
1928.05.17           1
1895.00.00.b    

Ahora voy a centrarme en limpiar por separado aquellas columnas que componen mi nueva tabla.

# A) Columna Case Number

En primer lugar, quiero eliminar las filas duplicadas en la columna Case Number. Esta columna es importante ya que nos da una estimación del número total de ataques de tiburón.

Después de haber observado el dataframe, he llegado a la conclusión de que cada una de las filas debería tener un valor diferente: vemos que cada registro está compuesto de la fecha del día del accidente (AAAA.MM.DD), a la que se le añade una letra en orden alfabético en los casos en los que ha habido más de un ataque en el día. 

Por supuesto, no tenemos la certeza de que los registros se nombren de esta manera. Aún mucho menos que quitando los duplicados, obtengamos el número real de ataques de tiburón. No obstante, teniendo tan poca información acerca de cómo se han recogido y plasmado los datos, ninguna metodología nos daría esa seguridad.

In [1135]:
#Eliminamos los registros duplicados si los hubiese
var1 = tiburon.drop_duplicates(subset=['Case Number'], keep='first')
tiburon['Case Number'] = var1
tiburon.shape


(25723, 3)

Vemos que no existían duplicados, ya que el número de filas se ha mantenido. Ahora vamos a proceder a cambiar el valor de los registros nulos por 0, ya que vamos a proceder a eliminarlos. 

In [1136]:
var2 = tiburon.dropna(how = 'any')
tiburon = var2
tiburon.shape

(5749, 3)

Ahora el dataframe con el que vamos a trabajar va a ser mucho más reducido, ya que hemos descendido de 25723 a 5749 filas.

## B) Columna Year

Elimino todos aquellos registros en los que su valor sea cero o nulo. También modifico el tipo de Year a un 'interger', para eliminar el '.0'.

In [1137]:
tiburon["Year"].isnull()
pd.to_numeric(tiburon["Year"], errors = "coerce")
tiburon = tiburon.dropna(subset = ["Year"])
tiburon["Year"] = tiburon["Year"].astype(int)

Ahora vamos a eliminar fechas inverosímiles, es decir, en este caso aquellas que sean inferiores a 1500.

In [1138]:
var3 = tiburon[(tiburon["Year"] >= 1500)]
tiburon = var3
tiburon.head()

Case Number  Year Fatal (Y/N)
0  2018.06.25  2018           N
1  2018.06.18  2018           N
2  2018.06.09  2018           N
3  2018.06.08  2018           N
4  2018.06.04  2018           N

## C) Columna Fatal (Y/N)

Además de lo hecho anteriormente, vamos a eliminar las filas en las que el valor en las columnas Year y Fatal (Y/N) sea igual a cero. Esto lo hacemos para que estos datos (que parecen incorrectos), no afecten a nuestros resultados.

In [1139]:
tiburon = tiburon[tiburon["Fatal (Y/N)"] != 0].copy()
tiburon.shape


(5622, 3)

Nuestra muestra ha descendido mucho, tal vez demasiado para que podamos sacar conclusiones significativas.

In [1140]:
tiburon.head()

Case Number  Year Fatal (Y/N)
0  2018.06.25  2018           N
1  2018.06.18  2018           N
2  2018.06.09  2018           N
3  2018.06.08  2018           N
4  2018.06.04  2018           N

Ahora vamos a homogeneizar los registros de manera a dejar tres opciones posibles (Y, N, Unknown). Como he encontrado muchos problemas al intentar resolver esta cuestión con replace y con regex, al final he optado por hacer un diccionario.

In [1141]:
def homog_fatal (v):

    fatal = {

        "Y":"Y", "N":"N", "UNKNOWN":"UNKNOWN","y":"Y", " Y":"Y", "Y ":"Y", "n":"N", "N ":"N", " N":"N","M":"UNKNOWN","2017":"UNKNOWN"

    }

    for fatal, fila in fatal.items(): 

        if fatal in v: 

            return fila

    return None

In [1142]:
tiburon["Fatal (Y/N)"] = tiburon["Fatal (Y/N)"].apply(homog_fatal)   
tiburon["Fatal (Y/N)"].value_counts()

N          4295
Y          1325
UNKNOWN       2
Name: Fatal (Y/N), dtype: int64

No obstante, como oodemos observar aquí, nos encontramos ante un error que no logro resolver. Solamente aparecen 2 filas con el nombre "UNKNOWN", a pesar de que anteriormente contábamos con 72 filas con ese valor, y que en el diccionario hemos incluído la posibilidad de que las filas tengan ese nombre.

In [1143]:
display(tiburon.head())

Case Number  Year Fatal (Y/N)
0  2018.06.25  2018           N
1  2018.06.18  2018           N
2  2018.06.09  2018           N
3  2018.06.08  2018           N
4  2018.06.04  2018           N

# Paso 3: Realizar operaciones estadísticas para averiguar si la hipótesis que hemos planteado se cumple.

En primer lugar, vamos a agrupar por rangos los años 1950 y los años 2000, que son las dos décadas que nos interesan para este estudio.

In [1144]:

var4 = tiburon[(1500 > tiburon["Year"] > 1600) & (2000 > tiburon["Year"] > 2010)
tiburon["Year"] = var4
tiburon
#var3 = 

SyntaxError: invalid syntax (<ipython-input-1144-bcfdac179238>, line 3)

In [ ]:
stats = tiburon.describe().transpose()
stats['IQR'] = stats['75%'] - stats['25%']
stats


In [ ]:
#No consigo hacer que el número de casos sea count
#tiburon['Case Number'] = [len(x) for x in tiburon ['Case number Count']]
#tiburon

In [ ]:
#Después de haber limpiado un poco mis columnas, voy a agrupar las columnas que me interesan
#de manera a poder estudiarlas con mayor profundidad

In [ ]:
#df_pivot = tiburon.pivot_table(values = 'len()', columns = 'Year','Fatal (Y/N)', index = ['Sex '])
#tiburon = df_pivot
#tiburon

In [ ]:
#var8 = tiburon.query('1950 <= Year <= 2010')
#tiburon = var8
#tiburon
#var9 = tiburon.query('2000 <= Year <= 2010')
#tiburon = var9
#tiburon

In [ ]:
print(tiburon.loc[(tiburon["Year"] > 1950) & (tiburon["Fatal (Y/N)"] == "Y") & (tiburon["Sex "] == "Y"),:])

In [ ]:
var8 = (tiburon[["Fatal (Y/N)","Sex "]].agg([pandas.Series.mean,pandas.Series.std])

#tiburon = var8
        
#tiburon